### PR TITLE
Add Multi-Module project, DataService, Mediator Project, Connector and .car file support

### PR DESCRIPTION
**Pull request by AVSSAM**

### CODE REVIEW

File: vscode-plugin/package.json [hunk 1]
```diff
@@ -169,12 +191,34 @@
...
+				"category": "WSO2EI"
+			},
+			{
+				"command": "wso2ei.dataservice.create.service",
+				"title": "Create New Data Service",
+				"category": "WSO2EI"
+			},
+			{
+				"command": "wso2ei.mediatorproject.create.project",
+				"title": "Create New Mediator project",
```
Comment (sajinieKavindya) [inline vscode-plugin/package.json:207]: Let's make the "p" in project capital here.

File: vscode-plugin/src/archetype/ArchetypeModule.ts [hunk 2]
```diff
@@ -37,7 +40,7 @@ export namespace ArchetypeModule {
      * Create new ESB Project from esb-project-archetype.
      */
     export async function createESBProject(): Promise<void> {
-        let artifactID: string | undefined = await showInputBoxForArtifactId();
+        let artifactID : string | undefined   = await showInputBoxForArtifactId();
```
Comment (sajinieKavindya) [inline vscode-plugin/src/archetype/ArchetypeModule.ts:43]: Let's fix the formatting issues here

File: vscode-plugin/src/archetype/ArchetypeModule.ts [hunk 3]
```diff
@@ -72,9 +76,48 @@ export namespace ArchetypeModule {
                 groupId: groupID,
                 artifactId: artifactID
             };
+
```
Comment (sajinieKavindya) [inline vscode-plugin/src/archetype/ArchetypeModule.ts:79]: Shall we remove the unwanted new lines in this file?

File: vscode-plugin/src/connector/ConnectorModule.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
+/*
+Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (sajinieKavindya) [inline vscode-plugin/src/connector/ConnectorModule.ts:2]: Update to the new license

File: vscode-plugin/package.json (hunk 1 above)
Comment (AVSSAM) [reply vscode-plugin/package.json]: fixed

File: vscode-plugin/src/archetype/ArchetypeModule.ts (hunk 2 above)
Comment (AVSSAM) [reply vscode-plugin/src/archetype/ArchetypeModule.ts]: fixed

File: vscode-plugin/src/archetype/ArchetypeModule.ts (hunk 3 above)
Comment (AVSSAM) [reply vscode-plugin/src/archetype/ArchetypeModule.ts]: fixed

File: vscode-plugin/src/connector/ConnectorModule.ts (hunk 4 above)
Comment (AVSSAM) [reply vscode-plugin/src/connector/ConnectorModule.ts]: license updated

**Comment by chamilaadhi** [comment]
@sajinieKavindya plz check and merge

File: vscode-plugin/package.json [hunk 5]
```diff
@@ -169,12 +207,64 @@
...
+				"category": "WSO2EI"
+			},
+			{
+				"command": "wso2ei.project.build",
+				"title": "Build Integration Project",
+				"category": "WSO2EI"
+			},
+			{
+				"command": "wso2ei.car.create.zip",
+				"title": "Build Integration Project From capp",
```
Comment (sajinieKavindya) [inline vscode-plugin/package.json:233]: capp -> CApp?

File: vscode-plugin/src/archetype/ArchetypeModule.ts [hunk 6]
```diff
@@ -40,24 +43,22 @@ export namespace ArchetypeModule {
...
         // Ensure that groupID name is valid.
-        while (typeof groupID !== "undefined" && !Utils.validate(groupID)) {
-            window.showErrorMessage("Enter valid GroupId name!!");
+        while (typeof groupID !== "undefined" && !Utils.validateGroupId(groupID)) {
+            window.showErrorMessage("Enter valid GroupId!!");
             groupID = await showInputBoxForGroupId();
         }
 
-        if (typeof artifactID === "undefined" || groupID === "undefined") {
+        if (typeof artifactID === "undefined" || typeof groupID === "undefined") {
```
Comment (sajinieKavindya) [inline vscode-plugin/src/archetype/ArchetypeModule.ts:58]: Can the type of the artifactID and groupID be ever "undefined" at this point as the above while loops execute until they are not "undefined"?

File: vscode-plugin/src/archetype/ArchetypeModule.ts [hunk 7]
```diff
@@ -72,9 +73,54 @@ export namespace ArchetypeModule {
...
+    export async function importProject() {
+        const homedir: string = require('os').homedir();
+        const targetFolderHint = Uri.file(homedir);
+
+        //get the target folder
+        const targetLocation: string | null = await chooseTargetFolder(targetFolderHint);
+
+        if (targetLocation) {
+
+            // let templateSettingsFilePath: string = path.join(dirName, '..', '..', 'templates', 'Conf', "settings.json");
```
Comment (sajinieKavindya) [inline vscode-plugin/src/archetype/ArchetypeModule.ts:97]: Can we remove these commented lines?

File: vscode-plugin/src/archetype/ArchetypeModule.ts [hunk 8]
```diff
@@ -72,9 +73,54 @@ export namespace ArchetypeModule {
...
+
+            commands.executeCommand('vscode.openFolder', Uri.file(targetLocation), true);
+            window.showInformationMessage("Project Imported Successfully");
+
+        }
+
+    }
 }
 
+
```
Comment (sajinieKavindya) [inline vscode-plugin/src/archetype/ArchetypeModule.ts:123]: Shall we remove these extra new lines?

File: vscode-plugin/src/archetype/Archetype.ts [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (sajinieKavindya) [inline vscode-plugin/src/archetype/Archetype.ts:2]: No need to change the year here. Normally, we do not change the year for a change of a file. Let's have the current year only for the new files in this PR.

File: vscode-plugin/src/archetype/archetypeUtils.ts [hunk 10]
```diff
@@ -18,6 +18,10 @@ Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 
 export const GROUP_ID_PREFIX: string = "com.example.";
 
-export const ARCHETYPE_ARTIFACT_ID: string = "org.wso2.carbon.extension.esb.project-archetype";
-export const ARCHETYPE_GROUP_ID: string = "org.wso2.carbon.extension.archetype";
-export const ARCHETYPE_VERSION: string = "2.0.16";
+//export const ARCHETYPE_ARTIFACT_ID: string = "org.wso2.carbon.extension.esb.project-archetype";
```
Comment (sajinieKavindya) [inline vscode-plugin/src/archetype/archetypeUtils.ts:21]: Shall we remove these commented lines?

File: vscode-plugin/src/archive/ArchiveModule.ts [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+
+let isEsbProjectCreated: boolean = false;
+let isRegistryResourcesProjectCreated: boolean = false;
+let isConnectorModuleCreated: boolean = false;
+let isDataServiceProjectCreated: boolean = false;
+
+export namespace ArchiveModule {
+
+    const dirName = __dirname;
+    const SRC = ESBArtifactPath.SRC;
```
Comment (sajinieKavindya) [inline vscode-plugin/src/archive/ArchiveModule.ts:50]: Do we need to assign the already defined constants to a separate set of contacts here? Shall we use the existing constants right away in this file, WDYT?

File: vscode-plugin/src/archive/ArchiveModule.ts [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+
+        //extract zip archive
+        let tmpDirectory: string = path.join(newProjectDirectory, "tmp");
+        fse.mkdirSync(tmpDirectory);
+        await extract(carFileLocation, { dir: tmpDirectory });
+
+        //read root artifact file
+        let rootArtifactsPath: string = path.join(tmpDirectory, "artifacts.xml");
+        if (!fse.existsSync(rootArtifactsPath)) {
+            window.showErrorMessage("Can not find root artifacts file.New project creation failed...!");
```
Comment (sajinieKavindya) [inline vscode-plugin/src/archive/ArchiveModule.ts:88]: Space between these two sentences?

File: vscode-plugin/src/archetype/ArchetypeModule.ts (hunk 6 above)
Comment (AVSSAM) [reply vscode-plugin/src/archetype/ArchetypeModule.ts:58]: if user clicks the editor, input box will be closed and those ids can be  "undefined"

File: vscode-plugin/package.json (hunk 5 above)
Comment (AVSSAM) [reply vscode-plugin/package.json]: fixed

File: vscode-plugin/src/archetype/ArchetypeModule.ts (hunk 8 above)
Comment (AVSSAM) [reply vscode-plugin/src/archetype/ArchetypeModule.ts]: fixed

File: vscode-plugin/src/archetype/archetypeUtils.ts (hunk 10 above)
Comment (AVSSAM) [reply vscode-plugin/src/archetype/archetypeUtils.ts]: fixed

File: vscode-plugin/src/archive/ArchiveModule.ts (hunk 11 above)
Comment (AVSSAM) [reply vscode-plugin/src/archive/ArchiveModule.ts:48]: fixed

File: vscode-plugin/src/archive/ArchiveModule.ts (hunk 12 above)
Comment (AVSSAM) [reply vscode-plugin/src/archive/ArchiveModule.ts]: fixed

File: vscode-plugin/src/archetype/Archetype.ts (hunk 9 above)
Comment (AVSSAM) [reply vscode-plugin/src/archetype/Archetype.ts]: fixed

File: vscode-plugin/src/archetype/ArchetypeModule.ts [hunk 13]
```diff
@@ -72,9 +72,33 @@ export namespace ArchetypeModule {
...
+        const homedir: string = require('os').homedir();
+        const targetFolderHint = Uri.file(homedir);
+
+        //get the target folder
+        const targetLocation: string | null = await chooseTargetFolder(targetFolderHint);
+
+        if (targetLocation) {
+
+            Utils.createVsCodeSettingsFile(targetLocation);
+
```
Comment (rosensilva) [inline vscode-plugin/src/archetype/ArchetypeModule.ts:97]: Shall we remove unnecessary white spaces and keep only where necessary

File: vscode-plugin/src/archive/archiveResolver.ts [hunk 14]
```diff
@@ -16,21 +16,170 @@ Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
...
+import { executeProjectBuildCommand } from "../mavenInternals/commandHandler";
+import { SubDirectories } from "../artifacts/artifactUtils";
+import { chooseTargetFolder, chooseTargetFile, showInputBox, showInputBoxForArtifactId, showInputBoxForGroupId } from "../utils/uiUtils";
+import { Utils } from "../utils/Utils";
+import { ArchiveModule } from "./ArchiveModule";
+
+var fileSystem = require('fs');
+var archiver = require('archiver');
+const extract = require('extract-zip');
+var AdmZip = require('adm-zip');
```
Comment (rosensilva) [inline vscode-plugin/src/archive/archiveResolver.ts:30]: Is there any reason to start the variable name in Caps? or shall we make it a camel case name?

**Review by rosensilva — APPROVED** [review]

**Pull request merged**